src/selection.ts

```diff
@@ -63,17 +66,39 @@ export class NodesSelection extends Selection {
   }
 }
 
+export class NodesSelection extends NodeSelection {
+  public $startNode: ResolvedPos
+  public $endNode: ResolvedPos
+
+  constructor($from: ResolvedPos, $to: ResolvedPos) {
+    super($from)
+    this.$startNode = $from
+    this.$endNode = $to
+  }
+
+  toJSON(): NodesSelectionJSON {
+    return {
+      type: 'nodes',
+      startNode: this.$startNode,
+      endNode: this.$endNode,
+    }
+  }
+}
+
 export const getSelectionChangeGroup = (state: ManuscriptEditorState) => {
   const selection = state.selection
   const $pos = isTextSelection(selection)
     ? selection.$cursor
-    : selection instanceof NodesSelection && selection.$from
+    : (selection instanceof NodesSelection ||
+        selection instanceof InlineNodesSelection) &&
+      selection.$from
   if ($pos) {
     return trackChangesPluginKey
       .getState(state)
       ?.changeSet.groupChanges.find(
         (changes) =>
-          changes.length > 1 &&
+          (changes.length > 1 ||
+            changes[0].dataTracked.operation === CHANGE_OPERATION.structure) &&
```

Comment:
any chance we can abstract somehow? can you please add inline comment to explain why structure change needs special treatment here?